packages/terminal/src/widget.ts

```diff
@@ -71,7 +71,9 @@ export class Terminal extends Widget implements ITerminal.ITerminal {
     this.title.label = this._trans.__('Terminal');
 
     session.messageReceived.connect(this._onMessage, this);
-    session.disposed.connect(this.dispose, this);
+    if (this.getOption('closeOnExit')) {
+      session.disposed.connect(this.dispose, this);
+    }
```

Comment:
If the settings is changed during a session, the new value is not taken into consideration. We need to look for the option when disposing the session.

```suggestion
    session.disposed.connect(
      () => {
        if (this.getOption('closeOnExit')) {
          this.dispose();
        }
      },
      this
    );
```

Reply:
Thanks!